novawallet/Modules/NetworkManagement/Traits/CustomNetworkSetupTrait.swift

```diff
@@ -0,0 +1,279 @@
+        let typeSetupWrapper = OperationCombiningService.compoundNonOptionalWrapper(
+            operationManager: OperationManager(operationQueue: operationQueue)
+        ) {
+            let partialChain = try partialChainWrapper
+                .targetOperation
+                .extractNoCancellableResultData()
+            
+            guard !partialChain.isEthereumBased else {
+                return CompoundOperationWrapper.createWithResult(partialChain)
+            }
+            
+            let rawRuntimeFetchWrapper = rawRuntimeFetchFactory.createMetadataFetchWrapper(
+                for: partialChain.chainId,
+                connection: connection
+            )
+            
+            let typeSetupOperation = ClosureOperation<PartialCustomChainModel> {
+                let rawRuntimeMedatada = try rawRuntimeFetchWrapper
+                    .targetOperation
+                    .extractNoCancellableResultData()
+                
+                let typeRegistrtryInfo = try typeRegistryFactory.createForMetadataAndDefaultTyping(
+                    chain: partialChain,
+                    runtimeMetadataItem: rawRuntimeMedatada
+                )
+                
+                let ethereumBased = typeRegistrtryInfo
+                    .typeRegistryCatalog
+                    .nodeMatches(
+                        closure: { ($0 as? FixedArrayNode)?.length == 20 },
+                        typeName: "Address",
+                        version: 0
+                    )
+                
+                return ethereumBased
+                    ? partialChain.adding([.ethereumBased])
+                    : partialChain
+            }
+            
+            typeSetupOperation.addDependency(rawRuntimeFetchWrapper.targetOperation)
+            
+            return CompoundOperationWrapper(
+                targetOperation: typeSetupOperation,
+                dependencies: rawRuntimeFetchWrapper.allOperations
+            )
+        }
+        
+        typeSetupWrapper.addDependency(wrapper: partialChainWrapper)
+        
+        return typeSetupWrapper
+    }
+    
+    func createMainAssetSetupWrapper(
+        for chain: PartialCustomChainModel,
+        systemPropertiesOperationFactory: SystemPropertiesOperationFactoryProtocol,
+        connection: ChainConnection
+    ) -> CompoundOperationWrapper<PartialCustomChainModel> {
+        let propertiesOperation: BaseOperation<SystemProperties>? = chain.isEthereumBased && chain.noSubstrateRuntime
+            ? nil
+            : systemPropertiesOperationFactory.createSystemPropertiesOperation(connection: connection)
+        
+        let assetOperation = ClosureOperation<PartialCustomChainModel> {
+            let properties = try propertiesOperation?.extractNoCancellableResultData()
+            let defaultEVMAssetPrecision: UInt16 = 18
+            let precision = properties?.tokenDecimals.first ?? defaultEVMAssetPrecision
```

Comment:
I would better separate the evm logic from substrate. Just to be sure we don't accedently set improper precision for substrate token

Reply:
done